lib/config/common.go

```diff
@@ -190,6 +193,68 @@ func (cfg *Server) ListenAndServe(srv *http.Server) error {
 	}
 }
 
+func (cfg *Server) Listen() (net.Listener, error) {
+	l, err := net.Listen("tcp", cfg.Addr)
+	if err != nil {
+		return nil, err
+	}
+
+	logrus.WithField("addr", cfg.Addr).Info("listening")
+	return l, nil
+}
+
+// Serve the HTTP requests on the specified listener, and gracefully close when the context is cancelled.
+func (cfg *Server) Serve(ctx context.Context, l net.Listener, srv *http.Server) (err error) {
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	eg.Go(func() error {
+		err := srv.Serve(l)
+		if err == http.ErrServerClosed {
+			return nil
+		}
+
+		return err
+	})
+
+	eg.Go(func() error {
+		select {
+		case <-ctx.Done():
+			logrus.Println("shutting down gracefully")
+			ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulTimeout())
+			defer cancel()
+			return srv.Shutdown(ctx)
+		case <-egCtx.Done():
+			return nil
+		}
+	})
+
+	return eg.Wait()
+}
+
+func (cfg *Server) GracefulTimeout() time.Duration {
+	if cfg.Graceful == 0 {
+		cfg.Graceful = DefaultGraceful
+	}
```

Comment:
Is there ever a case where we do actually want 0? Where there shouldn't be a graceful timeout? I imagine not but just checking

Reply:
Not actually changing this behaviour 🤷

Reply:
https://github.com/cuvva/cuvva-public-go/blob/master/lib/config/common.go#L161

Reply:
Fair :P